packages/init/node-async_hooks.js

```diff
@@ -0,0 +1,226 @@
+      bootstrapData.push(promiseData);
+    } else if (length) {
+      // console.warn('No candidates matched');
+    }
+  }
+
+  if (bootstrapData.length) {
+    // Normally all promise hooks are disabled in a subsequent microtask
+    // That means Node versions that modify promises at init will still
+    // trigger our proto hooks for promises created in this turn
+    // The following trick will disable the internal promise init hook
+    // However, only do this for destroy modifying versions, since some versions
+    // only modify promises if no destroy hook is requested, and do not correctly
+    // reset the internal init promise hook in those case. (e.g. v14.16.2)
+    const resetHook = createHook({});
+    resetHook.enable();
+    resetHook.disable();
+
+    const { asyncId, triggerAsyncId, resource } = bootstrapData.pop();
+    const symbols = Object.getOwnPropertySymbols(resource);
+    // const { length } = symbols;
+    let found = 0;
+    // if (length !== 3) {
+    //   console.error(`Found ${length} symbols on promise:`, ...symbols);
+    // }
+    symbols.forEach(symbol => {
+      const value = resource[symbol];
+      let type;
+      if (value === asyncId) {
+        type = 'async_id_symbol';
+      } else if (value === triggerAsyncId) {
+        type = 'trigger_async_id_symbol';
+      } else if (typeof value === 'object' && 'destroyed' in value) {
+        type = 'destroyed';
+      } else {
+        // console.error(`Unexpected symbol`, symbol);
+        return;
+      }
+
+      if (setAsyncSymbol(type, symbol)) {
+        found += 1;
+      }
+    });
+    return found;
+  } else {
+    // This node version is not mutating promises
+    return -2;
+  }
+};
+
+const getAsyncHookFallbackState = (promise, create) => {
+  let state = promiseAsyncHookFallbackStates.get(promise);
+  if (!state && create) {
+    state = {
+      [Symbol.nodeAsyncHooksAsyncId]: undefined,
+      [Symbol.nodeAsyncHooksTriggerAsyncId]: undefined,
+    };
+    if (Symbol.nodeAsyncHooksDestroyed) {
+      state[Symbol.nodeAsyncHooksDestroyed] = undefined;
+    }
+    promiseAsyncHookFallbackStates.set(promise, state);
+  }
+  return state;
+};
+
+const setAsyncIdFallback = (promise, symbol, value) => {
+  const state = getAsyncHookFallbackState(promise, true);
+
+  if (state[symbol]) {
+    if (state[symbol] !== value) {
+      // This can happen if a frozen promise created before hooks were enabled
+      // is used multiple times as a parent promise
+      // It's safe to ignore subsequent values
+    }
+  } else {
+    state[symbol] = value;
+  }
+};
+
+const getAsyncHookSymbolPromiseProtoDesc = (symbol, disallowGet) => ({
+  set(value) {
+    if (Object.isExtensible(this)) {
+      Object.defineProperty(this, symbol, {
+        value,
+        writable: false,
+        configurable: false,
+        enumerable: false,
+      });
+    } else {
+      // console.log('fallback set of async id', symbol, value, new Error().stack);
```

Comment:
tip. Especially if you're dealing with async_hooks, use `process._rawDebug` instead of `console.log/error`  
console API is triggering hooks, so if you end up using console in an async hook you get infinite loop.   
_rawDebug is the only really synchronous thing that writes output.

Reply:
That explains the infinite loop I got when I had a logging call in an `after` hook!